### PR TITLE
Add softmax and relevant element-wise binary and unary ops

### PR DESCRIPTION
**Pull request by huningxin**
Softmax is one op that is required by [the first-wave models](https://github.com/webmachinelearning/webnn/blob/master/op_compatibility/first_wave_models.md). Softmax is a kind of big op that can be composed by primitive `reduceSum`, `div` and `exp`. The `div` and `exp` are not in existing spec. This PR adds them.

According to the spec style and conventions introduced by PR #64, this PR also adds and groups the element-wise binary and unary ops in one go. These ops are supported by both ONNX and TensorFlow.

@anssiko @wchao1115 @gramalingam, please take a look. Thanks!

***
<a href="https://pr-preview.s3.amazonaws.com/huningxin/webnn/pull/68.html" title="Last updated on Jun 10, 2020, 8:08 AM UTC (d21cc16)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/webmachinelearning/webnn/68/12d9dfe...huningxin:d21cc16.html" title="Last updated on Jun 10, 2020, 8:08 AM UTC (d21cc16)">Diff</a>

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -654,6 +638,38 @@ partial interface NeuralNetworkContext {
...
+    input tensor.
+
+    <div class="note">
+    The behavior of this operation can be generically emulated from the usage of
+    other operations as follow. However, user agents typically have a more
+    efficient implementation for it, therefore its usage is encouraged from the
+    performance standpoint.
+    <pre highlight="js">
+    return nn.div(nn.exp(x), nn.reduceSum(nn.exp(x), [axis], /* keepDimensions = */ true));
+    </pre>
```
[the remark this answers is not in the document]
Comment (huningxin) [reply index.bs:668]: @wchao1115 , thanks for pointing out the incompatibility issue. The current definition actually follows the [TensorFlow `softmax`](https://www.tensorflow.org/api_docs/python/tf/nn/softmax) where the reduceSum of exponentials is only along the `axis`. However, as you mentioned, the `axis` of [ONNX `softmax`](https://github.com/onnx/onnx/blob/master/docs/Operators.md#Softmax) "describes the axis of the input when coerced to 2D" and the reduceSum of exponentials is along axis 1 of the coerced 2D tensor. So it seems to me that the common case is only when `axis` is the last dimension of the input tensor. On the other hand, the use case of softmax for first-wave models is only for 2D input tensor. With that, I propose to only support 2D input tensor and compute reduceSum of exponentials on axis 1 for `nn.softmax` for now. What do you think?

[the remark this answers is not in the document]
Comment (huningxin) [reply index.bs:668]: Another thing I need to fix is to subtract the the max value of x before compute the exponential.

File: index.bs [hunk 2]
```diff
@@ -654,6 +638,38 @@ partial interface NeuralNetworkContext {
     tensor is specified by the *newShape* argument.
 </div>
 
+### softmax ### {#api-neuralnetworkcontext-softmax}
+Compute the softmax (normalized exponential) values of the input tensor along a
+given axis.
```
[the remark this answers is not in the document]
Comment (huningxin) [reply index.bs:643]: Thanks for the pointer. I'll incorporate that.

File: index.bs [hunk 3]
```diff
@@ -249,22 +249,68 @@ interface NeuralNetworkContext {
 };
 </script>
 
-### add ### {#api-neuralnetworkcontext-add}
-Compute the element-wise binary addition of the two input tensors.
+### element-wise binary operations ### {#api-neuralnetworkcontext-binary}
```
[the remark this answers is not in the document]
Comment (huningxin) [reply index.bs:252]: Good suggestions. I'll fix them.

File: index.bs (hunk 1 above)
[the remark this answers is not in the document]
Comment (huningxin) [reply index.bs:665]: Pushed a new commit with support of the 2D input and fix of the implementation. Please take another look.

**Comment by huningxin** [comment]
BTW, @anssiko , I found the PR review bot didn't update the preview for this PR when I pushed new commits. And it even didn't create the preview for PR #73 . Do you know what happens?

**Comment by anssiko** [comment]
@tobie might know if there was an intermittent outage today in the otherwise amazing pr-preview.

**Comment by tobie** [comment]
That is weird.

**Comment by wchao1115** [comment]
@huningxin I had the same problem a couple days ago when I submitted my PR. The bot seems not working.

File: index.bs [hunk 4]
```diff
@@ -249,22 +249,68 @@ interface NeuralNetworkContext {
...
+<script type=idl>
+partial interface NeuralNetworkContext {
+  Operand abs(Operand x);
+  Operand ceil(Operand x);
+  Operand cos(Operand x);
+  Operand exp(Operand x);
+  Operand floor(Operand x);
+  Operand log(Operand x);
+  Operand neg(Operand x);
+  Operand sin(Operand x);
```
Comment (gramalingam) [inline index.bs:290]: Since we have sin and cos, should we have tan as well?

Comment (huningxin) [reply index.bs:341]: Add `tan` in the new commit. Please take a look.

**Comment by huningxin** [comment]
@wchao1115 and @gramalingam , I pushed new commits to address your comments. Please take another look. Thanks!

File: index.bs (hunk 1 above)
[the remark this answers is not in the document]
Comment (wchao1115) [reply index.bs:719]: @huningxin I can see why you want to compute the output values over the distances to the max value instead of the raw input values for numerical stability. Though this implementation trick is well known to some, it may still cause confusion to others when it doesn't seem to exactly match the canonical computation of the softmax values (even though the end result is the same mathematically) given that the description of this function also points to its canonical mathematical expression in Wikipedia.

So, my suggestion is to add some comment here to clarify *the different math* in your pseudo code, with something like "this sample deploys a well-known implementation trick to compute the exponentials of the distances to the max value, instead of the exponentials of the input values itself, in order to increase the numerical stability of the result".

I also find the explanation in the softmax paragraph of this [article](https://cs231n.github.io/linear-classify/#softmax) a worthwhile reference to this trick.

[the remark this answers is not in the document]
Comment (wchao1115) [reply index.bs:719]: @huningxin This is great. Please take a look at my inline comment on the updated pseudo code.

[the remark this answers is not in the document]
Comment (huningxin) [reply index.bs:730]: Great point. I added a new commit to incorporate your comment. Please take another look. Thanks!

**Comment by huningxin** [comment]
With the three approvals, I am going to merge this PR. Thanks for the reviews!
